txnkv/transaction/batch_getter.go

```diff
@@ -66,31 +66,67 @@ func NewBufferBatchGetter(buffer BatchBufferGetter, snapshot BatchGetter) *Buffe
 
 // BatchGet gets a batch of values.
 func (b *BufferBatchGetter) BatchGet(ctx context.Context, keys [][]byte) (map[string][]byte, error) {
-	if b.buffer.Len() == 0 {
+	bufferValues, err := b.buffer.BatchGet(ctx, keys)
+	if err != nil {
+		return nil, err
+	}
+	if len(bufferValues) == 0 {
 		return b.snapshot.BatchGet(ctx, keys)
 	}
-	bufferValues := make([][]byte, len(keys))
-	shrinkKeys := make([][]byte, 0, len(keys))
-	for i, key := range keys {
-		val, err := b.buffer.Get(ctx, key)
-		if err == nil {
-			bufferValues[i] = val
-			continue
+	shrinkKeys := make([][]byte, 0, len(keys)-len(bufferValues))
+	for _, key := range keys {
+		val, ok := bufferValues[string(key)]
+		if !ok {
+			shrinkKeys = append(shrinkKeys, key)
 		}
-		if !tikverr.IsErrNotFound(err) {
-			return nil, err
+		// the deleted key should be removed from the result, and also no need to snapshot read it again.
+		if len(val) == 0 {
+			delete(bufferValues, string(key))
```

Comment:
Should this `if` be put in an else branch of the above `if` statement?

We must explicitly state whether each BatchGet returns empty values.
For MemDBWithContext and PipelinedMemDB, they do.
For other `BatchGet`s, they probably don't?
Or shall we unify the behavior of the interface in a way that they all ignore empty values?

Reply:
Nice catch, this is a meaningless delete when `ok` is false.

`BatchGet` no more returns empty values in the current implementation, so we need to follow this rule.

https://github.com/tikv/client-go/blob/7bce6ea4dfa51c0ef98f16c9bbfc8c05144878b2/txnkv/transaction/batch_getter.go#L90-L92